apps/oc_chef_authz/src/oc_chef_policy_group_revision_association.erl

```diff
@@ -0,0 +1,207 @@
+%% -*- erlang-indent-level: 4;indent-tabs-mode: nil; fill-column: 92 -*-
+%% ex: ts=4 sw=4 et
+%% @author Daniel DeLeo <dan@chef.io>
+%% Copyright 2015 Chef Software, Inc. All Rights Reserved.
+%%
+%% This file is provided to you under the Apache License,
+%% Version 2.0 (the "License"); you may not use this file
+%% except in compliance with the License.  You may obtain
+%% a copy of the License at
+%%
+%%   http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing,
+%% software distributed under the License is distributed on an
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either express or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+
+-module(oc_chef_policy_group_revision_association).
+
+-include("../../include/oc_chef_types.hrl").
+
+-export([find_policy_revision_by_orgid_name_group_name/1,
+         insert_association/1,
+         update_association/1,
+         delete_association/1,
+
+         id/1,
+         fields_for_fetch/1,
+         fields_for_update/1,
+         find_query/0,
+         create_query/0,
+         update_query/0,
+         delete_query/0,
+         record_fields/0,
+         flatten/1,
+         is_indexed/0,
+         ejson_for_indexing/2]).
+
+find_policy_revision_by_orgid_name_group_name(Record) ->
+    case chef_sql:fetch_object(
+      chef_object:fields_for_fetch(Record),
+      element(1, Record),
+      chef_object:find_query(Record),
+      chef_object:record_fields(Record)
+      ) of
+        {ok, #oc_chef_policy_group_revision_association{} = Assoc} ->
+          Decompressed = decompress_record(Assoc),
+          {ok, Decompressed};
+        Any -> Any
+    end.
+
+insert_association(#oc_chef_policy_group_revision_association{} = Record) ->
+    ok = ensure_policy_exists(Record),
+    ok = ensure_policy_group_exists(Record),
+    ok = ensure_policy_revision_exists(Record),
```

Comment:
These do have FK constraints so they shouldn't be able to get in an incorrect state. What I'm attempting to do here is have code that's not in the WM resource module that can handle creating all the associated rows when needed so the WM resource just tells this module what it wants the final state of the DB to be. That said, I already need to fetch most of these items to do authz so by the time I get to this call I should know what things need to exist. I tried to find an example of how we do that sort of thing now, but the only thing that behaves like this is cookbooks+cookbook_versions, which is implemented as a special case in chef_db: https://github.com/chef/oc_erchef/blob/ed1eebbbc718e1c7dcfe7b28fb948b320adcadd1/apps/chef_db/src/chef_db.erl#L165-L167 eventually gets to https://github.com/chef/oc_erchef/blob/ed1eebbbc718e1c7dcfe7b28fb948b320adcadd1/apps/chef_db/src/chef_sql.erl#L1113-L1135

I'm open to suggestions on how that could be improved. Right now I'm thinking I could just add boolean fields in some place to indicate if the prerequisite objects exist.